[DI] Draft extension overrides docs (#21910)
backstage/backstage: docs/frontend-system/architecture/05-extension-overrides.md

## Introduction

An extension override is another building block in the Frontend System that allows you to programmatically override app or plugin extensions. Extension overrides are recommended in the following cases:

- Override the app skeleton layout;
- Override a default api implementation;
- Override a default app component;
- Override a plugin's default page;

> - Etc.

Reviewer: made into regular paragraph text

> It is also important to clarify when an override isn't needed. Creating an extension override is not recommended when:

Reviewer 2: I'm feeling that the real answer here is that you only need to override its implementation, and you should first make sure that you can't achieve the desired effect only through configuration instead. With the below being a few examples of how you might be able to configure an extension for your needs.

There is one interesting case to point out as well that is missing atm, and that is when the target extension has an input that allows you to configure it's behavior. You might for example have a page extension with an optional `header` input that you can use to override the default header component. In that case you would ideally create a plugin that provides the customized header extension, rather than overriding the entire page through an override.

Reviewer: made it into an example and explanatory text

- When you just want to configure a plugin, such as changing the attachment point, a path or any other setting. Be sure to read the extension documentation before overriding an extension to make sure you cannot configure it to work the way you want;

> - It is already possible to enable an extension that is already provided by the plugin. Imagine we have a Search plugin that provides a default search result item extension for rendering search results. Consider also that we have a Catalog plugin with a disabled Catalog search result item extension. To change the way Catalog search result items are rendered, you don't need to override the default result item extension. The only thing you need to do is enable the Catalog search result extension via the config file. For other types of search results, the default Search result item will continue to be used.

Reviewer 2: Feeling that this is perhaps a little it contrived 😅, isn't it basically "double check that there isn't already an extensions that does what you want that you could enable instead"?

Reviewer: made into regular paragraph text

## Creating an Extension Override

The following steps should be followed to override extensions:

```tsx
// plugins/aperture-overrides/src/overrides.ts
import {
  createThemeApi,
  createExtensionOverrides
} from '@backstage/frontend-plugin-api';

// Creating a light theme extension;

> const apertureLightTheme = createThemeApi({ … });

Reviewer 2: Thinking we might need to provide some more options here to show how to actually override the existing app themes

Author: Done 🙌🏻

  Extensions: [apertureLightTheme, apertureDarkTheme]
});

// plugins/aperture-overrides/src/index.ts
export { default } from './overrides';
```

For this example, we are creating overrides for the light and dark theme extensions and exporting the overrides from the plugin index file. Now we are able to use the overrides in a Backstage app:

```tsx
// packages/app/src/App.tsx
import { createApp } from '@backstage/frontend-app-api';
import apertureOverrides from ‘@backstage/plugin-aperture-overrides’

const app = createApp({
 features: [ apertureOverrides ],
});

export default app.createRoot().
```

If the plugin you want to change is internal to your company or you just want to replace one of the application's core extensions, you can decide to create replacements directly in the application for local replacements. See an example below:

```tsx
// packages/app/src/themes.ts
import { createExtensionOverrides } from from '@backstage/frontend-plugin-api';

// Creating the light theme extension;
export const apertureLightTheme = createThemeApi({ … });

// Creating the light theme extension;
const apertureDarkTheme = createThemeApi({ … });

// Exporting your custom extensions

> export default [apertureLightTheme, apertureDarkTheme];

Reviewer 2: I'm thinking we could just reference the package name that we export the override from in the first example instead? There's quite a lot of duplicate code here and I think we should encourage having overrides separated out into a different package.

Author: I replaced it with a single app example, is it better?

// packages/app/src/App.tsx
import { createApp } from '@backstage/frontend-app-api';
import themes from ‘./themes’

const app = createApp({
 features: [
   createExtensionOverrides({
     extensions: [
      ...themes,
     ],
   }),
 ],
});

export default app.createRoot();
```

> Note that it can still be a good idea to split your overrides out into separate packages in large projects. But it's up to you to decide how to group the extensions into extension overrides.

Reviewer 2: I think this could be replaced by the opposite: noting that it's recommended to split into a separate package but you can also keep the overrides directly in the app.

Author: Replaced ✅

## Overriding Existing Extensions

To override an existing extension (which is already provided by a plugin), you need to provide an extension that has the same ID as the existing extension. That is, all kind, namespaces, and name must match the extension you want to replace. This means that you typically need to provide an explicit namespace when overriding extensions from a plugin.

> Imagine that we have a plugin with search id and the plugin provides a search page extension that you want to fully override with your own custom component. To do so, you have to create your page extension and inform to the extension factory that the extension namespace is search to make sure you are overriding the search page of the plugin with id 'search' and not creating a anonymous page that will not override thesearch plugin one.

Reviewer 2: ```suggestion
Imagine you have a plugin with the ID `'search'`, and the plugin provides a page extension that you want to fully override with your own custom component. To do so, you need to create your page extension with an explicit `namespace` option that matches that of the plugin that you want to override, in this case `'search'`. If the existing extension also has an explicit `name` you'd need to set the `name` of your override extension to the same value as well.
```

```tsx
// packages/app/src/search.ts
const customSearchPage = createPageExtension({
  namespace: 'search',
  // Omitting name since it is the root plugin page
  defaultPath: 'search'
  loader: async () => <div>My custom search page</>
});
```

> We recommend that plugin developes share the extension ids in their plugin documentations, but usually you can infer the id by following the (Naming pattern)[./08-naming-patterns] standars.

Reviewer 2: ```suggestion
We recommend that plugin developes share the extension IDs in their plugin documentations, but usually you can infer the id by following the (Naming pattern)[./08-naming-patterns] standars.
```

### Example

In the example below, we create a file that exports custom extensions for the app's `light` and `dark` themes:

```tsx title="packages/app/src/themes.ts" {10-11,22-23} showLineNumbers
import {
  createThemeExtension,
  createExtensionOverrides
} from '@backstage/frontend-plugin-api';
import { apertureThemes } from './themes';
import { ApertureLightIcon, ApertureDarkIcon } from './icons';

// Creating a light theme extension
const apertureLightTheme = createThemeExtension({
  namespace: 'app',
  name: 'light',
  title: 'Aperture Light Theme',
  variant: 'light',
  icon: <ApertureLightIcon />,
  Provider: ({ children }) => (
    <UnifiedThemeProvider theme={apertureThemes.light} children={children} />
  ),
});

// Creating a dark theme extension
const apertureDarkTheme = createThemeExtension({
  namespace: 'app',
  name: 'dark',
  title: 'Aperture Dark Theme',
  variant: 'dark',
  icon: <ApertureDarkIcon />,
  Provider: ({ children }) => (
    <UnifiedThemeProvider theme={apertureThemes.dark} children={children} />
  ),
});

// Creating an extension overrides preset
export createExtensionOverrides({
  extensions: [apertureLightTheme, apertureDarkTheme]
});

> ```

Author: I think it is nice to group overrides by type (eg. all api overrides) or plugin (eg. all search plugin overrides), let's see if @Rugvip has any thoughts.

> ```

Reviewer 2: Just as an immediate thing: `app-next` is not our gold standard for app structure 😂. I think it makes sense to encourage separation into packages as much as possible, but ofc these can in the end be put directly in the app as well. Splitting it into separate modules for these examples feels like a sane middleground

> Lines `10` and `22` should be highlighted because they declare `namespace` as `'app'`, so the system knows we are overriding app extensions. Additionally, to specifically override the `light` and `dark` theme extensions, we set the `name` option to `light` and `dark` respectively on lines `11` and `23`. Therefore, to override app theme extensions, we ensure that the extension `namespace` and `name` match those of the default app theme extension definitions.

Reviewer 2: Line numbers are tricky to keep in sync, I think it would be best to avoid using the and instead refer to pieces of the example through text or `[N]` markers

Author: Done

Now we are able to use the overrides in a Backstage app:

```tsx title="packages/app/src/App.tsx" {5} showLineNumbers
import { createApp } from '@backstage/frontend-app-api';
import apertureOverrides from './themes';

const app = createApp({

>  features: [ apertureOverrides ],

Reviewer 2: perhaps `themeOverrides` or just `themes` instead?

Author: Done

```tsx title="packages/app/src/search.ts" showLineNumbers
import { createPageExtension } from '@backstage/frontend-plugin-api';

// Creating a custom search page extension
const customSearchPage = createPageExtension({
  namespace: 'search',
  // Omitting name since it is the index plugin page
  defaultPath: '/search',
  loader: () => import('./SearchPage').then(m => m.<SearchPage/>),
});

export createExtensionOverrides({
  extensions: [customSearchPage]
});
```

Don't forget to configure your overrides in the `createApp` function:

```tsx title="packages/app/src/App.tsx" {5} showLineNumbers
import { createApp } from '@backstage/frontend-app-api';
import searchOverrides from './search';

const app = createApp({
  features: [searchOverrides],
});

export default app.createRoot();
```

Now let's talk about the last override case, orphan extensions.

> ## Create Orphan Extensions

Reviewer 2: ```suggestion
## Create Standalone Extensions
```

Thinking this is perhaps a better term? "orphan" I think is perhaps a bit too easy to confuse with detached extensions that aren't part of the app tree?

Sometimes you just need to quickly create a new extension and not overwrite an app extension or plugin. You can also use overrides to create extensions, but remember that if you want to make this extension available for installation by other users, we recommend providing it via a plugin in a separate package.

### Example

> Imagine you want to create a page that is currently only used by your application, like an Institutional page for example, you can use overrides to extend the Backstage app to render it. To do so, simply create a page extension and pass it to the app as an override:

Reviewer 3: ```suggestion
Imagine you want to create a page that is currently only used by your application, like an Institutional page, for example. You can use overrides to extend the Backstage app to render it. To do so, simply create a page extension and pass it to the app as an override:
```

```tsx title="packages/app/src/App.ts" showLineNumbers
import { createApp } from '@backstage/frontend-app-api';
import {
  createPageExtension,
  createExtensionOverrides,
} from '@backstage/frontend-plugin-api';

const app = createApp({
  features: [
    createExtensionOverrides({
      extensions: [
        createPageExtension({
          name: 'institutional',
          defaultPath: '/institutional',
          loader: () =>
            import('./institutional').then(m => <m.InstitutionalPage />),
        }),
      ],
    }),
  ],
});

export default app.createRoot();
```

> Note that we are omitting `namespace` when creating the page extension. When we omit `namespace`, we are telling the system the new extension is orphaned (not an application or plugin extension) and this is all about orphaned extensions!

Reviewer 3: ```suggestion
Note that we are omitting `namespace` when creating the page extension. When we omit `namespace`, we are telling the system the new extension is orphaned and not an application or plugin extension!
```

## Introduction

An extension override is a building block of the Frontend System that allows you to programmatically override app or plugin extensions anywhere in your application. Since the entire application is built mostly out of extensions from the bottom up, this is a powerful feature. You can use it for example to provide your own app root layout, to replace the implementation of a Utility API with your a custom one, to override how the catalog page renders itself, and much more.

> Note that in general, most features should have a good level of customizability built into themselves, so that users do not have to leverage extension overrides to achieve common goals. A well written feature often has app-config settings, or uses extension inputs for extensibility where applicable. An example of this is the search plugin which allows you to provide result renderers as inputs rather than replacing the result page wholesale just to tweak how results are shown. Adopters should be taken advantage of those when possible, and only use extension overrides when it's necessary to entirely replace the extension. Check the respective extension documentation if available for guidance.

Reviewer: ```suggestion
Note that in general, most features should have a good level of customizability built into themselves, so that users do not have to leverage extension overrides to achieve common goals. A well written feature often has app-config settings, or uses extension inputs for extensibility where applicable. An example of this is the search plugin which allows you to provide result renderers as inputs rather than replacing the result page wholesale just to tweak how results are shown. Adopters should take advantage of those when possible, and only use extension overrides when it's necessary to entirely replace the extension. Check the respective extension documentation if available for guidance.
```

## Overriding App Extensions

If you want to override an app extension, you will need to create a new extension to replace the existing one and add it to the list of overridden features. The steps are:

1. Create your extension overrides

> In this example below, we are going to create custom extensions for the app light and dark themes:

Reviewer: In general, if you indent _everything_ by three spaces so that it's in line with the "Create your extension overrides" text above, then it'll be made part of the numbered list item as expected.

```
1. Foo

Bar
```

becomes 

%%%%%%%%%%%%
1. Foo

Bar

%%%%%%%%%%%%

but

```
1. Foo

   Bar
```

becomes

%%%%%%%%%%%%
1. Foo

   Bar

%%%%%%%%%%%%

Author: I restructured the page, thanks for the tip 🎉

Author: I also added a `Creating Orphan Extensions` sub-section, could you please take a look @freben ?

> ```tsx
> // packages/app/src/themes.ts

Reviewer: ```suggestion
```tsx title="packages/app/src/themes.ts"
```

this is a pretty nice trick that we use in a couple of places, see for example [the API migration docs](https://github.com/backstage/backstage/blob/63afa63eb8593009956e06fd51f1111ccb654bd0/docs/frontend-system/utility-apis/05-migrating.md?plain=1)

same in the other examples below

Note that it can still be a good idea to split your overrides out into separate packages in large projects. But it's up to you to decide how to group the extensions into extension overrides.

## Overriding Plugin Extensions

To override an extension that is provided by a plugin, you need to provide an new extension that has the same ID as the existing extension. That is, all kind, namespace, and name options must match the extension you want to replace. This means that you typically need to provide an explicit namespace when overriding extensions from a plugin.

Imagine you have a plugin with the ID `'search'`, and the plugin provides a page extension that you want to fully override with your own custom component. To do so, you need to create your page extension with an explicit `namespace` option that matches that of the plugin that you want to override, in this case `'search'`. If the existing extension also has an explicit `name` you'd need to set the `name` of your override extension to the same value as well.

```tsx
// packages/app/src/search.ts
const customSearchPage = createPageExtension({
  namespace: 'search',
  // Omitting name since it is the index plugin page

>   defaultPath: 'search',

Reviewer: ```suggestion
  defaultPath: '/search',
```

right?

>   loader: () => Promise.resolve(<div>My custom search page</div>),

Reviewer: ```suggestion
  loader: async () => <div>My custom search page</div>,
```

this is a matter of taste, feel free to ignore

Author: Replace with a `import(...)`
